File: h/QuickSort.java
public class Solution {
    public int[] quickSort(int[] array) {
    int low = 0;
    int high = array.length - 1;
    QuickSort (array, low, high);
    return array;
    }
  
    void QuickSort(int R[], int low, int high) {
      int i, j, temp;
      i = low;
      j = high;
      if (low < high) {
          temp = R[low];    //设置枢轴
          while (i != j) {
              while (j > i && R[j] >= temp) {
                  --j;
              }
              if (i < j) {
                  R[i] = R[j];
                  ++i;
              }
  
             while (i < j && R[i] < temp) {
                  ++i;
              }
              if (i < j) {
                  R[j] = R[i];
                  --j;
              }
          }
          R[i] = temp;
          QuickSort(R, low, i - 1);
          QuickSort(R, i + 1, high);
      }
  }
}

File: z/quick_sort.cc
// https://app.laicode.io/app/problem/10

class Solution {
 public:
  vector<int> quickSort(vector<int> array) {
    // write your solution here
    if (array.empty()) {
      return array;
    }
    quick_sort(array, 0, array.size() - 1);
    return array;
  }

  void quick_sort(vector<int>& array, int start, int end) {
    if (start >= end) {
      return;
    }
    int left = start;
    int right = end;
    int mid = start + (end - start) / 2;
    while (left <= right) {
      while (array[left] < array[mid]) {
        ++left;
      }
      while (array[right] > array[mid]) {
        --right;
      }
      if (left <= right) {
        swap(array[left++], array[right--]);
      }
    }
    quick_sort(array, start, right);
    quick_sort(array, left, end);
  }
};
